includes/class-wporg-gp-translation-events-stats-calculator.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+class WPORG_GP_Translation_Events_Stats_Row {
+	public int $created;
+	public int $reviewed;
+	public int $users;
+
+	public function __construct( $created, $reviewed, $users ) {
+		$this->created  = $created;
+		$this->reviewed = $reviewed;
+		$this->users    = $users;
+	}
+}
+
+class WPORG_GP_Translation_Events_Event_Stats {
+	/**
+	 * Associative array of rows, with the locale as key.
+	 *
+	 * @var WPORG_GP_Translation_Events_Stats_Row[]
+	 */
+	private array $rows = [];
+
+	private WPORG_GP_Translation_Events_Stats_Row $totals;
+
+	/**
+	 * @throws Exception
+	 */
+	public function add_row( string $locale, WPORG_GP_Translation_Events_Stats_Row $row ) {
+		if ( ! $locale ) {
+			throw new Exception( 'locale must not be empty' );
+		}
+		$this->rows[ $locale ] = $row;
+	}
+
+	public function set_totals( WPORG_GP_Translation_Events_Stats_Row $totals ) {
+		$this->totals = $totals;
+	}
+
+	/**
+	 * Get an associative array of rows, with the locale as key.
+	 *
+	 * @return WPORG_GP_Translation_Events_Stats_Row[]
+	 */
+	public function rows(): array {
+		return $this->rows;
+	}
+
+	public function totals(): WPORG_GP_Translation_Events_Stats_Row {
+		return $this->totals;
+	}
+}
+
+class WPORG_GP_Translation_Events_Stats_Calculator {
+	private string $ACTIONS_TABLE_NAME = WPORG_GP_Translation_Events_Translation_Listener::ACTIONS_TABLE_NAME;
+
+	/**
+	 * @throws Exception
+	 */
+	function for_event( WP_Post $event ): WPORG_GP_Translation_Events_Event_Stats {
+		$start = new DateTime( get_post_meta( $event->ID, '_event_start', true ), new DateTimeZone( 'UTC' ) );
+		$end   = new DateTime( get_post_meta( $event->ID, '_event_end', true ), new DateTimeZone( 'UTC' ) );
+		$stats = new WPORG_GP_Translation_Events_Event_Stats;
+		global $wpdb;
+
+		$query = $wpdb->prepare( "
+				select locale,
+					   sum(if(action = 'create', 1, 0)) AS created,
+					   sum(if(action in ('approve', 'reject', 'request_changes', 'mark_fuzzy'), 1, 0)) AS reviewed,
+					   count(distinct user_id) as users
+				from $this->ACTIONS_TABLE_NAME
+				where event_id = %d
+				  and happened_at between cast('%s' as datetime) and cast('%s' as datetime)
```

Comment:
Can we please drop this? Event entries should only go in the table when its the time of the event.

Reply:
I'm convinced this is needed. I've documented why that is in a a previous PR: https://github.com/WordPress/wporg-gp-translation-events/pull/3#discussion_r1462170267

Reply:
Ok, I read the comment and I think the solution needs to be to replace the `replace` with an `insert ignore`. We should not be updating the row.

Reply:
I'm having trouble understanding how the `insert ignore` would solve the problem, I think maybe I didn't explain what the problem is correctly.

Let's say I schedule WordCamp Lisbon for tomorrow. At 8am of the event's day a user adds a translation. However, it was raining that day so I decide to reschedule the event to next week.

Should the translation created by the user be included in the stats of the event?

Reply:
> what if we had a recurring Friday evening translations event?

This is a great point, it made me realise that indeed stats counting should be "dumb" and just use what is in the table, like you mention.

So I will remove `happened_at` from the table and adapt the query so it doesn't take it in account. I will also change `replace` to `insert ignore` when storing the action.

Reply:
Thank you!